File: src/org/elm/lang/core/parser/manual/IndentationHelper.java
package org.elm.lang.core.parser.manual;


import com.intellij.lang.PsiBuilder;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class IndentationHelper {
    private static final Pattern indentationPattern = Pattern.compile(".*[\r\n]([^\r\n]+)$", Pattern.DOTALL);

    public static int getIndentationOfPreviousToken(PsiBuilder builder) {
        // getTokenType has some side effects. Do not remove the call.
        builder.getTokenType();
        int end = builder.rawTokenTypeStart(0);
        return getLastIndentation(builder.getOriginalText().subSequence(0, end));
    }

    private static int getLastIndentation(CharSequence text) {
        Matcher m = indentationPattern.matcher(text);
        if (m.matches()) {
            return m.group(1).length();
        }
        return 0;
    }

    public static int getIndentation(CharSequence text, int start, int end) {
        CharSequence previousChars = text.subSequence(start, end);
        Matcher m = indentationPattern.matcher(previousChars);
        if (m.matches()) {
            return m.group(1).length();
        }
        return 0;
    }
}

File: src/main/kotlin/org/elm/lang/core/parser/manual/IndentationHelper.kt
package org.elm.lang.core.parser.manual


import com.intellij.lang.PsiBuilder
import java.util.regex.Pattern


internal object IndentationHelper {
    private val indentationPattern = Pattern.compile(".*[\r\n]([^\r\n]+)$", Pattern.DOTALL)

    fun getIndentationOfPreviousToken(builder: PsiBuilder): Int {
        // getTokenType has some side effects. Do not remove the call.
        builder.getTokenType()
        val end = builder.rawTokenTypeStart(0)
        return getIndentation(builder.originalText, 0, end)
    }

    fun getIndentation(text: CharSequence, start: Int, end: Int): Int {
        val previousChars = text.subSequence(start, end)
        val m = indentationPattern.matcher(previousChars)
        return if (m.matches()) m.group(1).length else 0
    }
}
